src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -154,195 +155,260 @@ protected override void DetachCore()
-
-                                    headers.Headers.Add(header.Header);
-                                    if ((header.HashBlock == getHeaders.HashStop) || (headers.Headers.Count == 2000))
-                                        break;
-                                }
-                            }
-                        }
+            this.logger.LogTrace("(-)");
+        }
 
-                        // Set our view of peer's tip equal to the last header that was sent to it.
-                        if (headers.Headers.Count != 0)
-                            this.pendingTip = this.Chain.GetBlock(headers.Headers.Last().GetHash()) ?? this.pendingTip;
+        /// <summary>
+        /// Processes "getheaders" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="getHeadersPayload">Payload of "getheaders" message to process.</param>
+        /// <remarks>
+        /// "getheaders" message is sent by the peer in response to "inv(block)" message 
+        /// after the connection is established, or in response to "headers" message
+        /// until an empty array is returned.
+        /// <para>
+        /// This payload notifies peers of our current best validated height,
+        /// which is held by consensus tip (not concurrent chain tip).
+        /// </para>
+        /// <para>
+        /// If the peer is behind/equal to our best height an empty array is sent back.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessGetHeadersAsync(NetworkPeer peer, GetHeadersPayload getHeadersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(getHeadersPayload), getHeadersPayload);
 
-                        peer.SendMessageVoidAsync(headers);
-                        break;
-                    }
+            if (!this.CanRespondToGetHeaders)
+            {
+                this.logger.LogTrace("(-)[CANT_RESPOND_TO_HEADERS]");
+                return;
+            }
+
+            // Ignoring "getheaders" from peers because node is in initial block download unless the peer is whitelisted.
+            if (this.initialBlockDownloadState.IsInitialBlockDownload() && !peer.Behavior<ConnectionManagerBehavior>().Whitelisted)
+            {
+                this.logger.LogTrace("(-)[IGNORE_ON_IBD]");
+                return;
+            }
+
+            HeadersPayload headers = new HeadersPayload();
+            ChainedBlock consensusTip = this.chainState.ConsensusTip;
+            consensusTip = this.Chain.GetBlock(consensusTip.HashBlock);
 
-                case HeadersPayload newHeaders:
+            ChainedBlock fork = this.Chain.FindFork(getHeadersPayload.BlockLocators);
+            if (fork != null)
+            {
+                if ((consensusTip == null) || (fork.Height > consensusTip.Height))
```

Comment:
Consider adding this to the above if condition as an "and" then you can remove the second ```if (fork != null)``` below

Reply:
old code